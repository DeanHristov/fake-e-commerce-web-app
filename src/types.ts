export enum PRODUCT_CATEGORIES {
  ELECTRONIC = 'ELECTRONIC',
}

export type TCurrency = 'USD' | 'BGN' | 'EUR';

export interface IComment {
  productId: string;
  _id: string;
  name: string;
  email: string;
  body: string;
  date: string;
}

// Object schema can be seen here: https://dummyjson.com/docs/products
export interface IProduct {
  id: number;
  title: string;
  description: string;
  price: number;
  discountPercentage: number;
  rating: number;
  stock: number;
  brand: string;
  category: string;
  thumbnail: string;
  images: string[];
}

export interface IUser {
  name: string;
  email: string;
  password: string;
  isAdmin: boolean;
}
