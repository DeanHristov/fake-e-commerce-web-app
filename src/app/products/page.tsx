import { FC } from 'react';
import { Metadata } from 'next';

export interface IProductsProps {}

export const metadata: Metadata = {
  title: 'Products',
};

const ProductsPage: FC<IProductsProps> = ({}) => {
  return <div className="h-[88vh]">Products page</div>;
};

export default ProductsPage;
