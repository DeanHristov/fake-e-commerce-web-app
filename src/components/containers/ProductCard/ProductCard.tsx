'use client';

import { FC, useCallback, useState } from 'react';
import { TCurrency } from '../../../types';
import { Utils } from '../../../utils/Utils';
import Button from '../../ui/Button';
import InputField from '../../ui/InputField';
import Widget from '../Widget';

export interface IProductCardProps {
  id: string;
  currency: TCurrency;
  countInStock: number;
  price: number;
}

const ProductCard: FC<IProductCardProps> = ({
  countInStock,
  currency,
  id,
  price,
}) => {
  const isAvailable: boolean = countInStock > 0;
  const [quantity, setQuantity] = useState<number>(0);
  const [cardStatus, setCardStatus] = useState();
  const onAddToCard = useCallback(() => {
    //TODO add logic here
  }, [quantity]);

  return (
    <div>
      <Widget>
        <div className="md:pb-2 flex justify-start items-center">
          <span className="md:p-2 w-[50%]  text-gray-500 text-2xl block">
            Price:
          </span>
          <span className="md:p-2 w-[50%]  text-gray-500 text-xl font-bold block">
            {Utils.parseAmountbyCurrency(price, currency)}
          </span>
        </div>
        <div className="md:pb-2 flex justify-start items-center  border-t border-gray-300">
          <span className="md:p-2 w-[50%] text-gray-500 text-2xl block">
            Status:
          </span>
          <span
            className={`md:p-2 w-[50%] block ${
              isAvailable ? 'text-green-400' : 'text-red-400'
            }`}
          >
            {isAvailable ? 'In stock' : 'Out of stock'}
          </span>
        </div>
        <div className="md:mb-8 flex justify-start items-center border-b border-t border-gray-300">
          <span className="md:p-2 w-[50%] text-gray-500 text-2xl block">
            Quantity:
          </span>
          <div className="w-[50%] md:h-10">
            <InputField onChange={(val) => setQuantity(Number(val))} />
          </div>
        </div>
        <Button
          disabled={!isAvailable}
          onClick={onAddToCard}
          title="Add to card"
        />
      </Widget>
    </div>
  );
};

export default ProductCard;
