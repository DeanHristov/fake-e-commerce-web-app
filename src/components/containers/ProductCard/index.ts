import ProductCard, { IProductCardProps } from './ProductCard';

export type { IProductCardProps };

export default ProductCard;
