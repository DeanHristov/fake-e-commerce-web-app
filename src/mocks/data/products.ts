import { IProduct, PRODUCT_CATEGORIES } from '../../types';

const products: IProduct[] = [
  {
    _id: '29a34b48-8343-4e3d-8c21-3548ba249936',
    name: 'Airpods Pro2 Headphones',
    image: '/images/airpods.jpg',
    description:
      "Apple's new H2 chip is the driving force behind AirPods Pro and their improved sound characteristics. It works in unison with a personalized driver and amplifier for crystal clear heights and deep, rich bass with incredible clarity, so every sound is more pronounced than ever",
    brand: 'Apple',
    category: PRODUCT_CATEGORIES.ELECTRONIC,
    price: 399.99,
    countInStock: 10,
    rating: 4.5,
    currency: 'USD',
    reviews: 5,
  },
];

export default products;
