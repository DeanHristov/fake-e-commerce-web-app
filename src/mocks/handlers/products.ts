import { MockedRequest, rest, RestContext, RestHandler } from 'msw';
import mockedProducts from '../data/products';
import { IProduct } from '@/types';

export let productsHandler: RestHandler<MockedRequest<IProduct>>[] = [
  // Fetching the all products
  rest.get(`${process.env.API_URL}/products`, (req, res, ctx: RestContext) => {
    return res(
      ctx.status(200),
      ctx.json({
        products: mockedProducts,
        total: 100,
        skip: 0,
        limit: 30,
      }),
    );
  }),

  // Fetching a single product
  rest.get(
    `${process.env.API_URL}/products/:productId`,
    (req, res, ctx: RestContext) => {
      const { productId } = req.params;
      const outputItem: IProduct | undefined = mockedProducts.find(
        (item: IProduct) => item.id === Number(productId),
      );

      return res(ctx.status(200), ctx.json(outputItem ?? null));
    },
  ),
];
