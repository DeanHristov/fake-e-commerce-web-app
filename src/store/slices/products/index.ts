export * from './productsSlice';
export * from './selectors';
