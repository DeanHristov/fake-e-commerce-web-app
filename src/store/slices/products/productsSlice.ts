import { IProduct } from '@/types';
import { Utils } from '@/utils/Utils';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

const reducerName = 'products';
const initialState: Record<string, IProduct> = Utils.tryToLoadFromStorage(
  reducerName,
  {},
);

export const productsSlice = createSlice({
  name: reducerName,
  initialState,
  reducers: {
    initProducts: (
      state: Record<string, IProduct>,
      action: PayloadAction<IProduct[]>,
    ) => {
      for (let productItem of action.payload) {
        state[productItem._id] = productItem;
      }
    },
  },
});

export const { initProducts } = productsSlice.actions;
export default productsSlice;
