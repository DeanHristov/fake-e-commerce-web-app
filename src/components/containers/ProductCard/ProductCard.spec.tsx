import { render } from '@testing-library/react';
import ProductCard from '.';
import { IProduct } from '@/types';
import products from '@/mocks/data/products';

const firstProduct: IProduct = products[0];

describe('Containers/ProductCard <ProductCard {...} />', () => {
  it('Should render a default state of the component', () => {
    const { container } = render(
      <ProductCard
        id={firstProduct.id}
        countInStock={firstProduct.stock}
        price={firstProduct.price}
        currency={'EUR'}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('Should render the component when countInStock=0', () => {
    const { container } = render(
      <ProductCard
        id={firstProduct.id}
        countInStock={firstProduct.stock}
        price={firstProduct.price}
        currency={'EUR'}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
