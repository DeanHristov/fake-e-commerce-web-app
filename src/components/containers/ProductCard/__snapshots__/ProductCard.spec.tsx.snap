// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers/ProductCard <ProductCard {...} /> Should render a default state of the component 1`] = `
<div>
  <div>
    <div
      class="bg-white border border-gray-300 rounded-lg p-4 shadow-lg"
    >
      <div
        class="py-0"
      />
      <div
        class="pt-2"
      >
        <div
          class="md:pb-2 flex justify-start items-center"
        >
          <span
            class="md:p-2 w-[50%]  text-gray-500 text-2xl block"
          >
            Price:
          </span>
          <span
            class="md:p-2 w-[50%]  text-gray-500 text-xl font-bold block"
          >
            €549.00
          </span>
        </div>
        <div
          class="md:pb-2 flex justify-start items-center  border-t border-gray-300"
        >
          <span
            class="md:p-2 w-[50%] text-gray-500 text-2xl block"
          >
            Status:
          </span>
          <span
            class="md:p-2 w-[50%] block text-green-400"
          >
            In stock
          </span>
        </div>
        <div
          class="md:mb-8 flex justify-start items-center border-b border-t border-gray-300"
        >
          <span
            class="md:p-2 w-[50%] text-gray-500 text-2xl block"
          >
            Quantity:
          </span>
          <div
            class="w-[50%] md:h-10"
          >
            <div
              class="relative w-full rounded-lg h-full border border-gray-500"
            >
              <input
                class="h-full px-6 w-full rounded-lg bg-white text-slate-700 ease-in duration-200 focus:bg-white"
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          class=""
        >
          <button
            class="cursor-pointer bg-green-600 w-full px-4 py-2 rounded-sm text-white font-bold uppercase active:shadow-md ease-in duration-200 flex justify-center items-center gap-2"
            type="button"
          >
            Add to card
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Containers/ProductCard <ProductCard {...} /> Should render the component when countInStock=0 1`] = `
<div>
  <div>
    <div
      class="bg-white border border-gray-300 rounded-lg p-4 shadow-lg"
    >
      <div
        class="py-0"
      />
      <div
        class="pt-2"
      >
        <div
          class="md:pb-2 flex justify-start items-center"
        >
          <span
            class="md:p-2 w-[50%]  text-gray-500 text-2xl block"
          >
            Price:
          </span>
          <span
            class="md:p-2 w-[50%]  text-gray-500 text-xl font-bold block"
          >
            €549.00
          </span>
        </div>
        <div
          class="md:pb-2 flex justify-start items-center  border-t border-gray-300"
        >
          <span
            class="md:p-2 w-[50%] text-gray-500 text-2xl block"
          >
            Status:
          </span>
          <span
            class="md:p-2 w-[50%] block text-green-400"
          >
            In stock
          </span>
        </div>
        <div
          class="md:mb-8 flex justify-start items-center border-b border-t border-gray-300"
        >
          <span
            class="md:p-2 w-[50%] text-gray-500 text-2xl block"
          >
            Quantity:
          </span>
          <div
            class="w-[50%] md:h-10"
          >
            <div
              class="relative w-full rounded-lg h-full border border-gray-500"
            >
              <input
                class="h-full px-6 w-full rounded-lg bg-white text-slate-700 ease-in duration-200 focus:bg-white"
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          class=""
        >
          <button
            class="cursor-pointer bg-green-600 w-full px-4 py-2 rounded-sm text-white font-bold uppercase active:shadow-md ease-in duration-200 flex justify-center items-center gap-2"
            type="button"
          >
            Add to card
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
